Multiclass Semantic Segmentation using DeepLabV3+

soumik12345 opened the pull request:
Semantic segmentation with the goal to assign semantic labels to every pixel in an image is one of the fundamental topics in computer vision. Deep convolutional neural networks based on the Fully Convolutional Neural Network show striking improvement over systems relying on hand-crafted features on benchmark tasks. In this example, we would implement
the **DeepLabV3+** model as proposed by the paper [Encoder-Decoder with Atrous Separable Convolution for Semantic Image Segmentation](https://arxiv.org/pdf/1802.02611.pdf) for multi-class semantic segmentation.

Please find Colab Notebook [here](https://colab.research.google.com/drive/1D_chASKoLqfneEzWQYGRbFLN-yvOyjiG?usp=sharing)

soumik12345 commented:
Hi @fchollet can you please review this example?

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+"""
+## Introduction
+"""
+
```
You can remove the break between the two text blocks (it's just one block)

soumik12345 replied:
Removed break between consecutive text blocks.

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+is one of the fundamental topics in computer vision. Deep convolutional neural networks
+based on the Fully Convolutional Neural Network show striking improvement over systems
+relying on hand-crafted features on benchmark tasks. In this example, we would implement
+the **DeepLabV3+** model as proposed by the paper [Encoder-Decoder with Atrous Separable
```
Make sure to keep markdown links on a single line, and generally follow [these guidelines](https://github.com/keras-team/keras-io/blob/master/contributor_guide.md)

soumik12345 replied:
Edited markdown links to be in single line.

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+
+"""shell
+!gdown https://drive.google.com/uc?id=1B9A9UCJYMwTL4oBEo4RZfbMZMaZhKJaz
+!unzip -q instance-level-human-parsing.zip
```
Remove the `!`, it gets added automatically

soumik12345 replied:
Removed `!`

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+        image = tf.image.decode_png(image, channels=1)
+        image.set_shape([None, None, 1])
+        image = tf.image.resize(images=image, size=[IMAGE_SIZE, IMAGE_SIZE])
+        image = tf.cast(image, tf.float32)
```
Is the cast necessary? The output of `resize` should be float32

soumik12345 replied:
Remove unnecessary cast.

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+    dataset = tf.data.Dataset.from_tensor_slices((image_list, mask_list))
+    dataset = dataset.map(load_data, num_parallel_calls=tf.data.AUTOTUNE)
+    dataset = dataset.batch(BATCH_SIZE, drop_remainder=True)
+    dataset = dataset.repeat(1)
```
Why?

soumik12345 replied:
Removed repeat and prefetch

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+"""
+
+"""
+**Atrous Convolution:** With Atrous convolution, as we go deeper, we can keep the stride
```
It's called "dilated convolution". "Convolution a trous" (literally "sparse convolution") is a misnomer.

soumik12345 replied:
Changed to dilated convolution.

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+    use_bias=False,
+):
+
+    x = tf.keras.layers.Conv2D(
```
You can import the `layers` module so you save a few characters each time you instantiate a layer

soumik12345 replied:
Added the following imports at the top and adjusted the layer, model, loss, and optimizer instantiations

```python3
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
```

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+
+def convolution_block(
+    block_input,
+    n_filters=256,
```
For consistency with API conventions, use `num_` as the number prefix

soumik12345 replied:
Replaced `n_filters` with `num_filters`

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+
+def AtrousSpatialPyramidPooling(aspp_input):
+
+    dims = tf.keras.backend.int_shape(aspp_input)
```
Just `aspp_input.shape`

soumik12345 replied:
Set `dims = aspp_input.shape`

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+
+    dims = tf.keras.backend.int_shape(aspp_input)
+
+    layer = tf.keras.layers.AveragePooling2D(pool_size=(dims[-3], dims[-2]))(aspp_input)
```
That's a tensor, not a layer, name it `x` or something

soumik12345 replied:
Renamed `layer` to `x`.

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+"""
+
+
+def DeeplabV3Plus():
```
You probably want to make `image_size` and `num_classes` signature arguments

soumik12345 replied:
Passed `image_size` and `num_classes` as arguments

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+use Adam as the optimizer.
+"""
+
+loss = tf.losses.SparseCategoricalCrossentropy(from_logits=True)
```
`keras.losses. SparseCategoricalCrossentropy`

soumik12345 replied:
Added the following imports at the top and adjusted the layer, model, loss, and optimizer instantiations

```python3
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
```

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+
+loss = tf.losses.SparseCategoricalCrossentropy(from_logits=True)
+model.compile(
+    optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
```
Likewise, import `keras` so you can do `keras.optimizers.Adam`

soumik12345 replied:
Added the following imports at the top and adjusted the layer, model, loss, and optimizer instantiations

```python3
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
```

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+    return predictions
+
+
+def decode_segmask(mask, colormap, n_classes):
```
decode_segmentation_masks -- in general use fully spelled out names everywhere

soumik12345 replied:
Renamed `decode_segmask` to `decode_segmentation_masks`

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,298 @@
+    model_output = tf.keras.layers.Conv2D(
+        NUM_CLASSES, kernel_size=(1, 1), padding='same')(layer)
+    
+    return tf.keras.Model(inputs=model_input, outputs=model_output)
```
`keras.Model`

soumik12345 replied:
Added the following imports at the top and adjusted the layer, model, loss, and optimizer instantiations

```python3
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
```

fchollet reviewed (changes requested):
Thanks for the PR! This is a great topic. I found the code quality to be very good -- great work!

soumik12345 commented:
Hi @fchollet thanks for the suggestions for improvement. I have updated the example as per your suggestions, can you please review?

fchollet commented on examples/vision/deeplabv3_plus.py:
```diff
@@ -0,0 +1,290 @@
+or the amount of computation. And finally, we can have larger output feature map which is
+good for semantic segmentation.
+
+The reason for using **Atrous Spatial Pyramid Pooling** is that it is discovered as the
```
Change "atrous" to "dilated" everywhere.

soumik12345 replied:
Changed atrous to dilated everywhere except the references.

fchollet reviewed (changes requested):
Thanks for the update! I pushed some copyedits. Please pull them first. Note that there are 2 TODOs left to fill in (text explanations).

soumik12345 commented:
Hi @fchollet I have updated the example as per your suggestions:

- Added a brief description of the CIHP dataset.
- Added a brief explanation of the inference and post-processing steps for visualization.
- Changed `AtrousSpatialPyramidPooling` to `DilatedSpatialPyramidPooling`.

Can you please review?

fchollet reviewed (commented):
Thanks for the update! I added a few minor copyedits. Please pull them, and after that you can add the generated files. The example looks great! 👍

soumik12345 commented:
Hi @fchollet , Added the generated files.

fchollet reviewed (approved):
Awesome, I'll take it from here. Thanks for the great contribution!

fchollet commented:
Upon closer review, I'd like to request one additional change (to be handled in a separate PR): please add a `validation_data` dataset in the call to `fit()`, and when visualizing the results, please use images from a validation or test set, rather than images from the training set. This is necessary to make sure that the model is capable of generalizing.

soumik12345 commented:
> Upon closer review, I'd like to request one additional change (to be handled in a separate PR): please add a `validation_data` dataset in the call to `fit()`, and when visualizing the results, please use images from a validation or test set, rather than images from the training set. This is necessary to make sure that the model is capable of generalizing.

Hi @fchollet , I raised a new PR updating the example as per your suggestions. Please find it at https://github.com/keras-team/keras-io/pull/629
